### ISSUE TITLE
JSNAPy failing in Ansible when multiple devices being tested

### ISSUE THREAD

**Issue by jnpr-bowen** [Juniper/jsnapy#263]
There seems to be some odd interaction between the JSNAPy logging code and how Ansible expects logging to work. I have developed an Ansible role to "package" testing into an "include role" statement, but there is something happening with logging in the JSNAPy module that breaks the Ansible task when run against multiple devices. It does not seem to appear if I force Ansible to be 'single threaded' (-f 1).
All of the JSNAPy configuration is default.
The error back from Ansible looks like:
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: 'NoneType' object has no attribute '__getitem__'
fatal: [lab-chi-lef01]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last): 
   File \"/tmp/ansible_fL3dow/ansible_module_junos_jsnapy.py\", line 179, in <module> 
     from jnpr.jsnapy import SnapAdmin 
   File \"/usr/local/lib/python2.7/dist-packages/jnpr/jsnapy/__init__.py\", line 70, in <module> 
     from jnpr.jsnapy.jsnapy import SnapAdmin 
   File \"/usr/local/lib/python2.7/dist-packages/jnpr/jsnapy/jsnapy.py\", line 35, in <module> 
     class SnapAdmin: 
   File \"/usr/local/lib/python2.7/dist-packages/jnpr/jsnapy/jsnapy.py\", line 38, in SnapAdmin 
     setup_logging.setup_logging() 
   File \"/usr/local/lib/python2.7/dist-packages/jnpr/jsnapy/setup_logging.py\", line 26, in setup_logging 
     for (handler, value) in iteritems(config['handlers']): 
 TypeError: 'NoneType' object has no attribute '__getitem__' 
 ", "module_stdout": "", "msg": "MODULE FAILURE"}

**Comment by sidhujasminder** [comment]
@jnpr-bowen it seems that the above error occurs if the logging.yml is not being read properly or it is empty somehow. Could you please check the logging.yml and jsnapy.cfg file and share the details of these files.

**Comment by jnpr-bowen** [comment]
I suspect it is somehow "not being read properly" due to some interaction with Ansible threading, because this happens to 'random' hosts only when I am testing multiple hosts in parallel. If I run ansible-playbook against one host or use "-f 1" to single thread Ansible, it does not happen. The playbook invocation looks like:
```
- name: Run all tests
  junos_jsnapy: 
    action: snapcheck
    callback: "{{reporting}}"
    host: "{{inventory_hostname}}"
    user: "{{username}}"
    passwd: "{{password}}"
    logfile: "{{jsnapy_test_base_dir}}/{{inventory_hostname}}/testing.log"
    port: 22
    test_files: "{{master_list}}"
  register: snap_res
```
The jsnapy config file is 'default'
```
# This file can be overwritten
# It contains default path for 
# config file, snapshots and testfiles
# If required, overwrite the path with your path
#config_file_path: path of main config file
#snapshot_path : path of snapshot file
#test_file_path: path of test file

[DEFAULT]
config_file_path= /etc/jsnapy
snapshot_path = /etc/jsnapy/snapshots
test_file_path = /etc/jsnapy/testfiles
```
The logging config file is 'default' as well
```disable_existing_loggers: true
formatters:
  custom_format:
    format: "%(asctime)s - %(name)s - %(levelname)s - %(hostname)s ............. \n\
      \ %(message)s"
  default:
    format: '%(message)s'
  default_file:
    format: "%(asctime)s - %(name)s - %(levelname)s - \n %(message)s"
  simple:
    format: '%(hostname)s-- %(message)s'
handlers:
  console:
    class: logging.StreamHandler
    formatter: default
    level: DEBUG
    stream: ext://sys.stdout
  critical_file_handler:
    backupCount: 20
    class: logging.handlers.RotatingFileHandler
    encoding: utf8
    filename: /var/log/jsnapy/jsnapy.log
    formatter: default_file
    level: CRITICAL
    maxBytes: 10485760
  debug_file_handler:
    backupCount: 20
    class: logging.handlers.RotatingFileHandler
    encoding: utf8
    filename: /var/log/jsnapy/jsnapy.log
    formatter: default_file
    level: DEBUG
    maxBytes: 10485760
  error_file_handler:
    backupCount: 20
    class: logging.handlers.RotatingFileHandler
    encoding: utf8
    filename: /var/log/jsnapy/jsnapy.log
    formatter: default_file
    level: ERROR
    maxBytes: 10485760
  info_file_handler:
    backupCount: 20
    class: logging.handlers.RotatingFileHandler
    encoding: utf8
    filename: /var/log/jsnapy/jsnapy.log
    formatter: default_file
    level: INFO
    maxBytes: 10485760
root:
  handlers:
  - console
  - debug_file_handler
  level: DEBUG
version: 1
```

**Comment by jnpr-bowen** [comment]
FYI: if you have already looked at the above comment, the first line of the logging.yml was not showing up. It got pasted onto the same line as the start comment ``` was on and so ended up not displaying. I have edited the comment to make it show up now.

**Comment by stacywsmith** [comment]
FYI, I ran into this issue while trying to test the Juniper.junos Ansible Galaxy modules before the next release. While this has only been seen with Ansible, this is purely a JSNAPy bug.

The issue is a race condition with reading the logging configuration file which happens in `setup_logging()`. `setup_logging()` is called when the `SnapAdmin` class is defined [here](https://github.com/Juniper/jsnapy/blob/master/lib/jnpr/jsnapy/jsnapy.py#L38), and that happens [here](https://github.com/Juniper/jsnapy/blob/master/lib/jnpr/jsnapy/__init__.py#L70) any time `jnpr.jsnapy` (or any sub-package) is imported.

The issue is that `setup_logging()` both [reads](https://github.com/Juniper/jsnapy/blob/master/lib/jnpr/jsnapy/setup_logging.py#L25) the log configuration file and [OVERWRITES](https://github.com/Juniper/jsnapy/blob/master/lib/jnpr/jsnapy/setup_logging.py#L39-L40)  the log configuration file. Actually, it overwrites the log configuration file once for every `handler` section in the log configuration file.

This means that if two processes attempt to import JSNAPy at roughly the same time (as Ansible does when `forks` is set to > 1, then there is a high probability that one process will attempt to read the configuration file while it is empty (in preparation to be re-written).

[This code](https://github.com/Juniper/jsnapy/blob/master/lib/jnpr/jsnapy/setup_logging.py#L26-L37) modifies the `filename` value of each `handler` section of the logging configuration file if running on Windows or in virtualenv. (It's worth noting that it uses a hard-coded path + filename which would totally overwrite the user's chosen path + filename, but that's a separate issue.)

The modified logging configuration data is then passed to the `logging` module [here](https://github.com/Juniper/jsnapy/blob/master/lib/jnpr/jsnapy/setup_logging.py#L41).

Before that happens, the modified logging configuration data is also written back to the logging configuration file [here](https://github.com/Juniper/jsnapy/blob/master/lib/jnpr/jsnapy/setup_logging.py#L39-L40) (by overwriting the file) with the parsed/modified logging configuration data.

I see no reason the modifications need to be written back to the logging configuration file (since the same modifications are applied each time anyway.)

I have tested by commenting out [lines 39 and 40](https://github.com/Juniper/jsnapy/blob/master/lib/jnpr/jsnapy/setup_logging.py#L39-L40) and this does indeed resolve the problem reported in this issue.

**Comment by sidhujasminder** [comment]
@jnpr-bowen @stacywsmith please try this new code and confirm if things are working fine.
